File: mobile/shared/src/commonTest/kotlin/uk/foodclassifier/ClassifierTest.kt
package uk.foodclassifier

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ClassifierTest {
    @Test
    fun veganIngredients() {
        val result = classifyIngredients("water, oats, sunflower oil, salt")
        assertEquals(DietClassification.Vegan, result.classification)
        assertFalse(result.hasWarnings)
    }

    @Test
    fun vegetarianIngredients() {
        val result = classifyIngredients("flour, milk, sugar, butter")
        assertEquals(DietClassification.Vegetarian, result.classification)
        assertTrue(result.matchedNonVegan.contains("milk"))
    }

    @Test
    fun nonVegetarianIngredients() {
        val result = classifyIngredients("tomatoes, chicken stock, basil")
        assertEquals(DietClassification.Neither, result.classification)
        assertTrue(result.matchedNonVegetarian.contains("chicken"))
    }

    @Test
    fun parmigianoReggianoIsUncertain() {
        val result = classifyIngredients("Parmigiano Reggiano Cheese (Cows' Milk)")
        assertEquals(DietClassification.Uncertain, result.classification)
        assertEquals(DietClassification.Vegetarian, result.definitiveClassification)
    }

    @Test
    fun bovineCollagenIsNeither() {
        val result = classifyIngredients("Bovine Collagen Hydrolysate, milk")
        assertEquals(DietClassification.Neither, result.classification)
        assertTrue(result.matchedNonVegetarian.contains("bovine"))
    }

    @Test
    fun parentheticalMilkIsDetected() {
        val result = classifyIngredients("Whey Protein Isolate (MILK), cocoa")
        assertEquals(DietClassification.Vegetarian, result.classification)
        assertTrue(result.matchedNonVegan.contains("milk"))
    }

    @Test
    fun parentheticalMilkOnlyIsDetected() {
        val result = classifyIngredients("Protein Blend (Calcium Caseinate (Milk))")
        assertEquals(DietClassification.Vegetarian, result.classification)
        assertTrue(result.matchedNonVegan.contains("milk"))
    }

    @Test
    fun squareBracketMilkIsDetected() {
        val result = classifyIngredients(
            "Protein Blend [Calcium Caseinate (Milk), Whey Protein Isolate (Milk)]",
        )
        assertEquals(DietClassification.Vegetarian, result.classification)
        assertTrue(result.matchedNonVegan.contains("milk"))
    }

    @Test
    fun ocrStyleMilkIsDetected() {
        val result = classifyIngredients(
            "Protein Blend [Calcium Caseinate Mi ME-NGR, Whey Protein Isolate lMilk]",
        )
        assertEquals(DietClassification.Vegetarian, result.classification)
        assertTrue(result.matchedNonVegan.contains("milk"))
    }

    @Test
    fun softCheeseOcrExtractClassifiesAsVegetarian() {
        val ocrText =
            "REDUCED FAT SOFT CHEESE NUTRITION:Typical v 632kJ/152kcal. Fat 10 " +
                "(arbohydrate 3.8a, of Protein 95a, Sat 070 Energy 190k/46kcal " +
                "NGREDIENTS: Reduced Fat Soft C(hese (Milk), Salt Stablisers: " +
                "Guar Gum, Carrageenan; Citrus Fibre"

        val parsed = uk.foodclassifier.text.IngredientLabelParser.parse(ocrText)
        val result = classifyIngredients(parsed.chips)
        assertEquals(DietClassification.Vegetarian, result.classification)
        assertTrue(
            result.matchedNonVegan.any { it in setOf("milk", "cheese") },
        )
    }
}

File: flutter/test/classifier_test.dart
import 'package:food_classifier/food_classifier.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('Classifier', () {
    test('veganIngredients', () {
      final result = classifyIngredients('water, oats, sunflower oil, salt');
      expect(result.classification, DietClassification.vegan);
      expect(result.hasWarnings, isFalse);
    });

    test('vegetarianIngredients', () {
      final result = classifyIngredients('flour, milk, sugar, butter');
      expect(result.classification, DietClassification.vegetarian);
      expect(result.matchedNonVegan, contains('milk'));
    });

    test('nonVegetarianIngredients', () {
      final result = classifyIngredients('tomatoes, chicken stock, basil');
      expect(result.classification, DietClassification.neither);
      expect(result.matchedNonVegetarian, contains('chicken'));
    });

    test('parmigianoReggianoIsUncertain', () {
      final result =
          classifyIngredients("Parmigiano Reggiano Cheese (Cows' Milk)");
      expect(result.classification, DietClassification.uncertain);
      expect(result.definitiveClassification, DietClassification.vegetarian);
    });

    test('bovineCollagenIsNeither', () {
      final result =
          classifyIngredients('Bovine Collagen Hydrolysate, milk');
      expect(result.classification, DietClassification.neither);
      expect(result.matchedNonVegetarian, contains('bovine'));
    });

    test('parentheticalMilkIsDetected', () {
      final result = classifyIngredients('Whey Protein Isolate (MILK), cocoa');
      expect(result.classification, DietClassification.vegetarian);
      expect(result.matchedNonVegan, contains('milk'));
    });

    test('parentheticalMilkOnlyIsDetected', () {
      final result =
          classifyIngredients('Protein Blend (Calcium Caseinate (Milk))');
      expect(result.classification, DietClassification.vegetarian);
      expect(result.matchedNonVegan, contains('milk'));
    });

    test('squareBracketMilkIsDetected', () {
      final result = classifyIngredients(
        'Protein Blend [Calcium Caseinate (Milk), Whey Protein Isolate (Milk)]',
      );
      expect(result.classification, DietClassification.vegetarian);
      expect(result.matchedNonVegan, contains('milk'));
    });

    test('ocrStyleMilkIsDetected', () {
      final result = classifyIngredients(
        'Protein Blend [Calcium Caseinate Mi ME-NGR, Whey Protein Isolate lMilk]',
      );
      expect(result.classification, DietClassification.vegetarian);
      expect(result.matchedNonVegan, contains('milk'));
    });
  });

  group('ClassificationMessages', () {
    test('describesDairyInPlainLanguage', () {
      final messages = describeClassificationMatches([], ['milk', 'whey']);
      expect(messages, hasLength(1));
      expect(messages.first, contains("cow's milk or other dairy"));
    });

    test('describesBovineInPlainLanguage', () {
      final messages =
          describeClassificationMatches(['bovine', 'collagen'], []);
      expect(messages, hasLength(1));
      expect(messages.first, contains('derived from cattle'));
    });

    test('describesMilkAndBovineSeparately', () {
      final messages = describeClassificationMatches(
        ['bovine'],
        ['milk'],
      );
      expect(messages, hasLength(2));
      expect(messages.any((message) => message.contains('cattle')), isTrue);
      expect(messages.any((message) => message.contains('dairy')), isTrue);
    });

    test('describesChickenInPlainLanguage', () {
      final messages = describeClassificationMatches(['chicken'], []);
      expect(messages, hasLength(1));
      expect(messages.first, contains('poultry'));
    });

    test('describesLikelyVeganWhenUncertain', () {
      final message = describeUncertainResult(DietClassification.vegan);
      expect(message, contains('likely vegan'));
      expect(message, contains('Check the label'));
    });

    test('describesLikelyVegetarianWhenUncertain', () {
      final message = describeUncertainResult(DietClassification.vegetarian);
      expect(message, contains('likely vegetarian'));
      expect(message, contains('Check the label'));
    });
  });
}
